src/frontend/src/components/authenticateBox.ts

```diff
@@ -0,0 +1,227 @@
+        unreachable(result);
+        break;
+    }
+  };
+
+  return new Promise<LoginData>((resolve) => {
+    authenticateBoxTemplate({
+      templates,
+      addDevice: () => addRemoteDevice(connection),
+      onContinue: async (userNumber) => {
+        const loginData = await authenticate(connection, userNumber).then(
+          retryOnError
+        );
+        setUserNumber(loginData.userNumber);
+        resolve(loginData);
+      },
+      register: async () => {
+        const loginData = await registerIfAllowed(connection).then(
+          retryOnError
+        );
+        setUserNumber(loginData.userNumber);
+        resolve(loginData);
+      },
+      recoverAnchor: (userNumber) => useRecovery(connection, userNumber),
+      userNumber: getUserNumber(),
+    });
+  });
+};
+
+export const authenticateBoxTemplate = (props: PageProps): void => {
+  const container = document.getElementById("pageContent") as HTMLElement;
+  render(pageContent(props), container);
+};
+
+const pageContent = ({
+  templates,
+  addDevice,
+  onContinue,
+  recoverAnchor,
+  register,
+  userNumber,
+}: PageProps): TemplateResult => {
+  const anchorInput = mkAnchorInput({
+    userNumber,
+    onSubmit: onContinue,
+  });
+
+  return html` <div class="l-container c-card c-card--highlight">
+      <!-- The title is hidden but used for accessibility -->
+      <h1 class="is-hidden">Internet Identity</h1>
+      <div class="c-logo">${icLogo}</div>
+      <div class="l-stack t-centered">
+        ${templates.message} ${templates.chasm ? mkChasm(templates.chasm) : ""}
+      </div>
+
+      ${anchorInput.template}
+
+      <button
+        @click="${anchorInput.submit}"
+        id="authorizeButton"
+        class="c-button"
+      >
+        Authorize
+      </button>
+      ${mkLinks({
+        register,
+        recoverAnchor: () => recoverAnchor(anchorInput.readUserNumber()),
+        addDevice,
+      })}
+    </div>
+    ${footer}`;
+};
+
+export const authenticate = async (
+  connection: Connection,
+  userNumber: bigint
+): Promise<LoginFlowSuccess | LoginFlowError> => {
+  try {
+    const result = await withLoader(() => connection.login(userNumber));
+    return apiResultToLoginFlowResult(result);
+  } catch (error) {
+    return {
+      tag: "err",
+      title: "Authentication Failed",
+      message: "An error occurred during authentication.",
+      detail: error instanceof Error ? error.message : JSON.stringify(error),
+    };
+  }
+};
+
+const mkChasm = ({ info, message }: ChasmOpts): TemplateResult => {
+  /* the chasm that opens to reveal details about alternative origin */
+  const chasmRef: Ref<HTMLDivElement> = createRef();
+
+  /* the (purely visual) arrow on the chasm */
+  const chasmToggleRef: Ref<HTMLSpanElement> = createRef();
+
+  /* Toggle the chasm open/closed */
+  const chasmToggle = () =>
+    withRef(chasmRef, (chasm) => {
+      const classes = chasm.classList;
+
+      if (classes.contains("c-chasm--closed")) {
```

Comment:
not in this pr but would be neat to have a utils/helper to toggle classes I guess